crates/sui-light-client/src/main.rs

```diff
@@ -0,0 +1,714 @@
+    for ckp_id in &checkpoints_list.checkpoints {
+        // check if there is a file with this name ckp_id.yaml in the checkpoint_summary_dir
+        let mut checkpoint_path = config.checkpoint_summary_dir.clone();
+        checkpoint_path.push(format!("{}.yaml", ckp_id));
+
+        // If file exists read the file otherwise download it from the server
+        let summary = if checkpoint_path.exists() {
+            read_checkpoint(config, *ckp_id)?
+        } else {
+            // Download the checkpoint from the server
+            let summary = download_checkpoint_summary(config, *ckp_id).await?;
+            summary.clone().verify(&prev_committee)?;
+            // Write the checkpoint summary to a file
+            write_checkpoint(config, &summary)?;
+            summary
+        };
+
+        // Print the id of the checkpoint and the epoch number
+        println!(
+            "Epoch: {} Checkpoint ID: {}",
+            summary.epoch(),
+            summary.digest()
+        );
+
+        // Extract the new committee information
+        if let Some(EndOfEpochData {
+            next_epoch_committee,
+            ..
+        }) = &summary.end_of_epoch_data
+        {
+            let next_committee = next_epoch_committee.iter().cloned().collect();
+            prev_committee = Committee::new(summary.epoch().saturating_add(1), next_committee);
+        } else {
+            return Err(anyhow!(
+                "Expected all checkpoints to be end-of-epoch checkpoints"
+            ));
+        }
+    }
+
+    Ok(())
+}
+
+async fn read_full_checkpoint(checkpoint_path: &PathBuf) -> anyhow::Result<CheckpointData> {
+    let mut reader = fs::File::open(checkpoint_path.clone())?;
+    let metadata = fs::metadata(checkpoint_path)?;
+    let mut buffer = vec![0; metadata.len() as usize];
+    reader.read_exact(&mut buffer)?;
+    bcs::from_bytes(&buffer).map_err(|_| anyhow!("Unable to parse checkpoint file"))
+}
+
+async fn write_full_checkpoint(
+    checkpoint_path: &Path,
+    checkpoint: &CheckpointData,
+) -> anyhow::Result<()> {
+    let mut writer = fs::File::create(checkpoint_path)?;
+    let bytes = bcs::to_bytes(&checkpoint)
+        .map_err(|_| anyhow!("Unable to serialize checkpoint summary"))?;
+    writer.write_all(&bytes)?;
+    Ok(())
+}
+
+async fn get_full_checkpoint(config: &Config, seq: u64) -> anyhow::Result<CheckpointData> {
+    let mut checkpoint_path = config.checkpoint_summary_dir.clone();
+    checkpoint_path.push("untrusted_cache");
+    checkpoint_path.push(format!("{}.bcs", seq));
+
+    // Try reading the cache
+    if let Ok(checkpoint) = read_full_checkpoint(&checkpoint_path).await {
+        return Ok(checkpoint);
+    }
+
+    // Downloading the checkpoint from the server
+    let client: Client = Client::new(config.rest_url());
+    let full_check_point = client.get_full_checkpoint(seq).await?;
+
+    // Add to cache
+    write_full_checkpoint(&checkpoint_path, &full_check_point).await?;
```

Comment:
The full_check_point data received is not verified, caching it might lead to caching fabricated checkpoint data? Or is it assumed that the cache is not trusted anyway?

Reply:
Removing cache to avoid this issue.